New errorfmt for Rust

fuine opened the pull request:
This PR closes #592. Stable 1.12 version of Rust, which introduces new errorfmt as default, will land in a few days and i think that it's reasonable to switch to the new errorfmt.

Introduced makers try to remove unnecessary informations, while maintaining error/warning messages, error description taken from the pointer line and note/help lines. The only thing that is missing is proper parsing of error/warning width, as pointer line contains both markers (^^^^) and error description.

More information about new format can be found [here](https://blog.rust-lang.org/2016/08/10/Shape-of-errors-to-come.html). I decided against parsing lifetime messages, because in my opinion they can't be shown in a sane way.

blueyed commented:
Thanks!

Do you feel like fixing the vint issues in the touched files?  (just add abort to all functions, i.e. `function! neomake#makers#cargo#cargo() abort`)

(Ref https://github.com/Kuniwak/vint/issues/181)

I think we should not merge this before the new version is released, should we?

fuine commented:
Sure, I will fix vints issues for rust. I'm not sure if we shouldn't leave old errorfmt for the time being and just add new to makers. People might still use older versions of compiler for some time and maybe it'd not be the worst idea. That said i have almost no experience with errorfmts and neomake and i'm not sure if long list of errorfmts doesn't induce some kind of performance hit.

blueyed commented:
Please check https://github.com/neomake/neomake/issues/549#issuecomment-250550544.

YorickPeterse commented on autoload/neomake/makers/cargo.vim:
```diff
@@ -1,11 +1,16 @@
+            \ '%Eerror: %m,'.
+            \ '%Wwarning: %m,'.
+            \ '%Inote: %m,'.
+            \ '%-Z\ %#-->\ %f:%l:%c,'.
```
@fuine I believe `-Z` is deprecated, and it will trigger warnings. For example, using this Rust file:

```rust
fn main() {
    let $syntax_error = "hello";
}
```

The output in the quickfix window is:

```
|| the option `Z` is unstable and should only be used on the nightly compiler, but it is currently accepted for backwards compatibility; this will soon change, see issue #31847 for more details
/tmp/test.rs|2 col 9 error| expected pattern, found `$`
|| aborting due to 2 previous errors
```

YorickPeterse replied:
See https://github.com/rust-lang/rust/issues/31847 for more information.

fuine replied:
Are you referring to the -Z no-trans in rustc maker? I should probably remove that flag as it is currently unstable. I think that cargo maker is fine.

YorickPeterse replied:
@fuine I think the entire `-Z` flag is unstable at the moment, but I'm not certain.

fuine replied:
@YorickPeterse i am aware of that, but the only usage of that flag in this PR is in the `rust.vim` file. I would like to know why did you comment on `cargo.vim` file. Is there something i am missing here?

YorickPeterse replied:
@fuine I commented on this particular line because it was the first instance of `-Z` being used, that's all.

fuine replied:
Ah ok, this particular -Z is actually the end of multiline message in vim's errorfmt (refer to [docs](http://vimdoc.sourceforge.net/htmldoc/quickfix.html#errorformat) to see more. Anyway thanks for pointing that out, i removed it from rust.vim.

fuine commented:
@antoyo i decided against putting `= note` and `= help` in the errorfmt because the only place i have seen them were clippy error messages and it looks like they don't come in one chunk (neomake interprets error/warning message as one chunk and then slurps additional `= note/help` lines as another chunk and we lose information about line/file at this point. If you've seen other examples where this is not the case feel free to describe it and we will try to put this in the errorfmt.

@YorickPeterse i removed `-Zno-trans` from the rustc call. Now neomake will just call rustc without any flags which will build whole file and emit binary. I'm not sure if it's desired behaviour, but i wasn't able to find any other stable flags that would prevent emission of binary and only report errors/warns. One solution would be to keep on using `-Zno-trans` and ignore the warning in the errorfmt, but i don't think that is a good solution.

@blueyed I think the PR is ready to be merged. One thing that is interesting is the json error format which currently is unstable but will probably stabilize soon. Do you have any plans to support such format? As you can see [here](https://i.imgur.com/XfO6vXN.png) it contains a lot more precise information to work with (locaction list contains current output of the new errorfmt and on the right you can see the 'normal' error message and json output). Currently there's no sane way to parse the lifetime annotations and show them in neomake.

blueyed commented:
> neomake interprets error/warning message as one chunk and then slurps additional = note/help lines as another chunk and we lose information about line/file at this point

Sounds like it's related to buffer_output (https://github.com/neomake/neomake/issues/598)?!  Or do you mean something else?

As for json: there is `json_decode` and we could map it to location list entries I guess?!
See https://github.com/neomake/neomake/issues/428.

fuine commented:
> > neomake interprets error/warning message as one chunk and then slurps additional = note/help lines  as another chunk and we lose information about line/file at this point

> Sounds like it's related to buffer_output (#598)?! Or do you mean something else?

Hmm, i had the impression that it chopped the output of the clippy (one consistent warning message was parsed as 2 lines of output), but i have just tried it again and this time it works. I'm not sure if it was related to stderr flushing by clippy or maybe it was me who made some stupid mistake. Anyways it works now and i have added additional `help` annotations that sometimes appear in the compiler output. 

I noticed that we use the same errorfmt in three different files (`clippy.vim`, `cargo.vim`, `rust.vim`). Wouldn't it make more sense if we define it in one place and then use the definition in these three makers?

As for json i will look into it after that error format stabilizes. I think that we should merge this PR first and if it will be more beneficial to parse json then new PR should be opened.

blueyed commented:
> Hmm, i had the impression that it chopped the output of the clippy (one consistent warning message was parsed as 2 lines of output), but i have just tried it again and this time it works. 

It depends on when it's flushed and e.g. if it is received during the "1st second" (but with only second-based resolution) - very fragile code there.. working on it in https://github.com/neomake/neomake/pull/607.

> I noticed that we use the same errorfmt in three different files (clippy.vim, cargo.vim, rust.vim). Wouldn't it make more sense if we define it in one place and then use the definition in these three makers?

Totally..! You should use the existing/leading/main function then to just return it in the other maker definitions.
Do you want to do this in this PR?  Should be a separate commit then I'd say.

blueyed commented:
JFI: #598 has been merged - maybe it helps?! (i.e. please rebase)

fuine commented:
I have rebased and introduced a single definition of rust's errorfmt (in `rust.vim`). Please note that i have almost no experience in writing vimscript so any feedback on that small refactor would be helpful.
I have also decided to add ignore patterns against 3 error messages that i feel are redundant.

Do you want me to squash my commits before you merge this PR?

blueyed commented on autoload/neomake/makers/ft/rust.vim:
```diff
@@ -1,19 +1,24 @@
+            \ '%G\ %#|\ %#%\\^%\\+ %m,'.
+            \ '%I%>help:\ %#%m,'.
+            \ '%Z\ %#%m,'.
+            \ '%-G%s,',
```
Not sure about the trailing comma here (in the errorformat), better remove it if it's not there on purpose.

blueyed commented:
LGTM, please squash and address the comment I've left.

fuine commented:
Removed the trailing comma and squashed, feel free to merge.

blueyed commented:
Thanks a lot!

jonhoo commented:
@fuine I'm not too familiar with the internals of neomake, so I don't know how hard this would be to do, but currently most type errors are simply rendered as `cargo: mismatched types`. This is not a particularly helpful error message. I suspect the reason is that Rust's new error messages include the type information on *later* lines of the output, and not in the `error:` line. Is there some way to still match the type information and emit it in the neomake output?

fuine commented:
@jonhoo you are right, but the maker matches these lines. If you open quickfix/loclist you will see the rest of the error. There's no easy way to make it appear by deafult without opening lists.

jonhoo commented:
That's true, but running `:cope` and `:cclo` (even with hotkeys) is sort of tedious. If we could come up with a good regex to bring the types in question into the statusline, that would be awesome! I don't even know where to begin implementing something like that though :/

fuine commented:
I agree, can you open an issue about that problem? I will try to do something about it in the nearest future, because that is one of the few inconveniences that i hit with the current state of things.

jonhoo commented:
@fuine see #793.
